### ISSUE TITLE
collectd crashes

### ISSUE THREAD

**Issue by stephen-soltesz** [m-lab/collectd-mlab#14]
collectd appears to crash on mlab2.nuq1t.

Two problems:

 * collectd is never restarted.
 * the export script runs unconditionally, which generates reports for all metrics as zero.

Why is collectd crashing?
Add watchdog to restart if collectd is not running? How to handle missing data? How to distinguish between monitoring failure vs system failures?

**Comment by stephen-soltesz** [comment]
collectd was generating a "Bus error" on mlab2.nuq1t and exiting immediately.

 * Set `/etc/default/collectd:DAEMON_COREFILE_LIMIT=unlimited`
 * See trace from gdb below.

Command name abbreviations are allowed if unambiguous.
(gdb) bt
#0  0x00330df9 in __memcpy_ssse3 () from /lib/libc.so.6
#1  0x0011e467 in rrd_write () from /usr/lib/librrd_th.so.4
#2  0x00123c26 in ?? () from /usr/lib/librrd_th.so.4
#3  0x001240dd in ?? () from /usr/lib/librrd_th.so.4
#4  0x001244ab in rrd_update_r () from /usr/lib/librrd_th.so.4
#5  0x00111213 in ?? () from /usr/lib/collectd/rrdtool.so
#6  0x00111aa7 in ?? () from /usr/lib/collectd/rrdtool.so
#7  0x00723b39 in start_thread () from /lib/libpthread.so.0
#8  0x002dec2e in clone () from /lib/libc.so.6

**Comment by stephen-soltesz** [comment]
The node where this was running had a bad disk so I suspected this. But, I also erased the rrd files under `/var/lib/collectd/rrd` wondering if one might be corrupted and collectd has been running since, about 24hrs.

Unfortunately, I erased the files that were corrupted. It would be helpful to have had them available to create a check to notice and prevent this kind of issue from recurring.

Simply restarting collectd will not work if the RRD files are corrupted, because it will crash again.

**Comment by stephen-soltesz** [comment]
A possible hint for detecting corrupted rrds: http://mailman.verplant.org/pipermail/collectd/2009-January/002339.html

The brute force approach might also be to auto-restart collectd, make a note of this, and if there are more than N restarts over 10 minutes, remove all old rrd files so collectd can recreate them.

**Comment by stephen-soltesz** [comment]
To determine if collectd has crashed we might be able to use more specific core patterns.
e.g. https://www.kernel.org/doc/Documentation/sysctl/kernel.txt and search for "core_pattern".

        /etc/sysctl.cfg:
        kernel.core_pattern = /tmp/core-%e-%s-%u-%g-%p-%t

**Comment by stephen-soltesz** [comment]
The crashes recurred, which gave me an opportunity to test how to detect this kind of error.

The hint link above suggested using "rrdtool fetch", but this didn't help in my case. I discovered that "rrdtool dump" would report problematic files. But this is a much more expensive operation. So, it should be run rarely, e.g. only if collectd has crashed.

        find ./ -name "*.rrd" | while read FILE ; do
              rrdtool dump "$FILE" > /dev/null || echo "$FILE" ;
        done

Also, the following information should be logged when a corruption is discovered: name of file, when it was fixed. This may provide better hints for tracking how often this occurs, and help track down root causes or find ideas about fixing the issue in a better way.
